### PR TITLE
You can now hide grenades in plushies

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
add: You can now add grenades to plushies! You'll need to cut out the stuffing first.
/:cl:

You can now add grenades to plushies. You need to cut out some of the stuffing with something sharp to make room for the grenade, but once you do you can simply slip any grenade in there to conceal a deadly surprise in the plushie. Petting the plushie will prime any grenades that can be activated directly in-hand, but has to be activated in another way if it has assemblies like a signaler/igniter.

You can remove grenades from plushies with wirecutters, and examining them from up-close will let you see if they have a grenade in them.

### CODE REVIEW

**Comment by GunHog** [comment]
You are a horrible person.

**Comment by MoreRobustThanYou** [comment]
yes.

**Comment by darkpaladin109** [comment]
What plushes do we have anyway? The only ones I'm aware of are the carp plushies.

**Comment by QualityVan** [comment]
i kinda feel like it should be any object of sufficiently small size but this's good too

**Comment by Xhuis** [comment]
@darkpaladin109:
* Slime
* Lizard
* Snek
* Nuke op
* Bubblegum
* Ratvar
* Nar-Sie

I know you can get some via toy crate, but not which ones.

File: code/game/objects/items/plushes.dm [hunk 1]
```diff
@@ -7,14 +7,64 @@
...
-	to_chat(user, "<span class='notice'>You pet [src]. D'awww.</span>")
+	if(stuffed || grenade)
+		to_chat(user, "<span class='notice'>You pet [src]. D'awww.</span>")
+		if(grenade && !grenade.active)
+			if(istype(grenade, /obj/item/grenade/chem_grenade))
+				var/obj/item/grenade/chem_grenade/G = grenade
+				if(G.nadeassembly) //We're activated through different methods
+					return
+			playsound(user, 'sound/weapons/armbomb.ogg', 10, TRUE)
+			addtimer(CALLBACK(grenade, /obj/item/grenade/proc/prime), grenade.det_time)
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: These two lines should be a /grenade proc

File: code/game/objects/items/plushes.dm [hunk 2]
```diff
@@ -7,14 +7,64 @@
...
 	resistance_flags = FLAMMABLE
 	var/list/squeak_override //Weighted list; If you want your plush to have different squeak sounds use this
+	var/stuffed = TRUE //If the plushie has stuffing in it
+	var/obj/item/grenade/grenade //You can remove the stuffing from a plushie and add a grenade to it for *nefarious uses*
 
 /obj/item/toy/plush/Initialize()
 	. = ..()
 	AddComponent(/datum/component/squeak, squeak_override)
 
 /obj/item/toy/plush/attack_self(mob/user)
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: Add a Destroy() which QDEL_NULLs grenade

File: code/game/objects/items/plushes.dm [hunk 3]
```diff
@@ -7,14 +7,64 @@
...
+			playsound(user, 'sound/weapons/armbomb.ogg', 10, TRUE)
+			addtimer(CALLBACK(grenade, /obj/item/grenade/proc/prime), grenade.det_time)
+	else
+		to_chat(user, "<span class='notice'>You try to pet [src], but it has no stuffing. Aww...</span>")
+
+/obj/item/toy/plush/attackby(obj/item/I, mob/living/user, params)
+	if(istype(I, /obj/item/wirecutters))
+		if(!grenade)
+			if(!stuffed)
+				to_chat(user, "<span class='danger'>You already murdered it!</span>")
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: warning span

File: code/game/objects/items/plushes.dm [hunk 4]
```diff
@@ -7,14 +7,64 @@
...
+				return
+			user.visible_message("<span class='notice'>[user] tears out the stuffing from [src]!</span>", "<span class='notice'>You rip a bunch of the stuffing from [src]. Murderer.</span>")
+			playsound(user, 'sound/items/wirecutter.ogg', 50, TRUE)
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: playsound(I, I.usesound, ...

File: code/game/objects/items/plushes.dm [hunk 5]
```diff
@@ -7,14 +7,64 @@
...
+			stuffed = FALSE
+		else
+			to_chat(user, "<span class='notice'>You remove the grenade from [src].</span>")
+			grenade.forceMove(get_turf(user))
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: remove this line, put_in_hands takes care of it

File: code/game/objects/items/plushes.dm [hunk 6]
```diff
@@ -7,14 +7,64 @@
...
+		if(stuffed)
+			to_chat(user, "<span class='danger'>You need to remove some stuffing first!</span>")
+			return
+		if(grenade)
+			to_chat(user, "<span class='danger'>[src] already has a grenade!</span>")
+			return
+		user.visible_message("<span class='warning'>[user] slides [grenade] into [src].</span>", \
+		"<span class='danger'>You slide [I] into [src].</span>")
+		grenade = I
+		user.transferItemToLoc(I, src)
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: this may return FALSE, you need to handle that

File: code/game/objects/items/plushes.dm [hunk 7]
```diff
@@ -7,14 +7,64 @@
...
+			stuffed = FALSE
+		else
+			to_chat(user, "<span class='notice'>You remove the grenade from [src].</span>")
+			grenade.forceMove(get_turf(user))
+			user.put_in_hands(grenade)
+			grenade = null
+		return
+	if(istype(I, /obj/item/grenade))
+		if(stuffed)
+			to_chat(user, "<span class='danger'>You need to remove some stuffing first!</span>")
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: warning span

File: code/game/objects/items/plushes.dm [hunk 8]
```diff
@@ -7,14 +7,64 @@
...
+			return
+		if(grenade)
+			to_chat(user, "<span class='danger'>[src] already has a grenade!</span>")
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: warning span

File: code/game/objects/items/plushes.dm [hunk 9]
```diff
@@ -7,14 +7,64 @@
...
+		user.transferItemToLoc(I, src)
+		return
+	. = ..()
+
+/obj/item/toy/plush/examine(mob/user)
+	..()
+	if(user.Adjacent(src))
+		if(!stuffed)
+			if(grenade)
+				to_chat(user, "<span class='warning'>[grenade] has been hidden in its body!</span>")
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: Too powergamey imo, let it be a surprise

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Needs a handle_atom_del to clear the grenade reference

File: code/game/objects/items/plushes.dm (hunk 9 above)
Comment (Xhuis) [reply code/game/objects/items/plushes.dm]: 💣 😱 💥

File: code/game/objects/items/plushes.dm (hunk 1 above)
Comment (Xhuis) [reply code/game/objects/items/plushes.dm]: What do you mean by that? Should the `/obj/item/grenade` typepath be replaced with something like `grenade.proc/prime`?

File: code/game/objects/items/plushes.dm [hunk 10]
```diff
@@ -23,49 +27,39 @@
...
 /obj/item/toy/plush/attackby(obj/item/I, mob/living/user, params)
 	if(istype(I, /obj/item/wirecutters))
 		if(!grenade)
 			if(!stuffed)
-				to_chat(user, "<span class='danger'>You already murdered it!</span>")
+				to_chat(user, "<span class='warning'>You already murdered it!</span>")
 				return
 			user.visible_message("<span class='notice'>[user] tears out the stuffing from [src]!</span>", "<span class='notice'>You rip a bunch of the stuffing from [src]. Murderer.</span>")
-			playsound(user, 'sound/items/wirecutter.ogg', 50, TRUE)
+			playsound(user, I.usesound, 50, TRUE)
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: Change user to I

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Still needs a handle_atom_del

**Comment by MoreRobustThanYou** [comment]
@Xhuis make it so you can also use knives / scalpels / etc to clear out plushies

**Comment by Cyberboss** [comment]
Yeah, just use the IS_SHARP flag

**Comment by Bawhoppen** [comment]
Oh good, people already asked for what I wanted too

File: code/game/objects/items/plushes.dm [hunk 11]
```diff
@@ -7,14 +7,58 @@
...
 	. = ..()
-	to_chat(user, "<span class='notice'>You pet [src]. D'awww.</span>")
+	if(stuffed || grenade)
+		to_chat(user, "<span class='notice'>You pet [src]. D'awww.</span>")
+		if(grenade && !grenade.active)
+			if(istype(grenade, /obj/item/grenade/chem_grenade))
+				var/obj/item/grenade/chem_grenade/G = grenade
+				if(G.nadeassembly) //We're activated through different methods
+					return
+			grenade.preprime(volume = 10)
```
Comment (kevinz000) [inline code/game/objects/items/plushes.dm]: should log

File: code/game/objects/items/plushes.dm [hunk 12]
```diff
@@ -7,14 +7,58 @@
...
+			to_chat(user, "<span class='warning'>You need to remove some stuffing first!</span>")
+			return
+		if(grenade)
+			to_chat(user, "<span class='warning'>[src] already has a grenade!</span>")
+			return
+		if(!user.transferItemToLoc(I, src))
+			return
+		user.visible_message("<span class='warning'>[user] slides [grenade] into [src].</span>", \
+		"<span class='danger'>You slide [I] into [src].</span>")
+		grenade = I
```
Comment (kevinz000) [inline code/game/objects/items/plushes.dm]: should log

File: code/game/objects/items/plushes.dm [hunk 13]
```diff
@@ -18,6 +18,10 @@
 	QDEL_NULL(grenade)
 	. = ..()
 
+/obj/item/toy/plush/handle_atom_del(atom/A)
+	if(A == grenade)
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: Call ..()

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by fludd12** [comment]
Is it possible to genericize this to any small-ish item? Hiding a shiv in a plushie to get a nerd out of prison, for example.

File: code/game/objects/items/plushes.dm (hunk 12 above)
Comment (PKPenguin321) [reply code/game/objects/items/plushes.dm]: No, if you can take it in and out (which you can) it will be a potential spam exploit

**Comment by Xhuis** [comment]
It's probably possible, yeah. For now I'm fine with keeping it like this, but in a later PR I might generalize it.

**Comment by Cyberboss** [comment]
s/generalize/componentize

File: code/game/objects/items/grenades/grenade.dm [hunk 14]
```diff
@@ -52,18 +52,19 @@
...
+/obj/item/grenade/proc/preprime(mob/user, delayoverride, msg = TRUE, volume = 60)
 	var/turf/T = get_turf(src)
-	log_grenade(user, T)
-	if(iscarbon(user))
-		var/mob/living/carbon/C = user
-		C.throw_mode_on()
-	if(msg)
-		to_chat(user, "<span class='warning'>You prime \the [src]! [det_time/10] seconds!</span>")
-	playsound(loc, 'sound/weapons/armbomb.ogg', 60, 1)
+	if(user)
```
Comment (AnturK) [inline code/game/objects/items/grenades/grenade.dm]: Should still log location if there's no user.

**Review by AnturK — DISMISSED** [review]

File: code/game/objects/items/grenades/grenade.dm [hunk 15]
```diff
@@ -52,18 +52,19 @@
...
-	playsound(loc, 'sound/weapons/armbomb.ogg', 60, 1)
+	log_grenade(user, T) //Inbuilt admin procs already handle null users
+	if(user)
+		add_fingerprint(user)
+		if(iscarbon(user))
+			var/mob/living/carbon/C = user
+			C.throw_mode_on()
+		if(msg)
+			to_chat(user, "<span class='warning'>You prime \the [src]! [det_time/10] seconds!</span>")
+	playsound(loc, 'sound/weapons/armbomb.ogg', volume, 1)
```
Comment (Cyberboss) [inline code/game/objects/items/grenades/grenade.dm]: playsound(src, ...

File: code/game/objects/items/plushes.dm [hunk 16]
```diff
@@ -7,14 +7,66 @@
...
+	var/stuffed = TRUE //If the plushie has stuffing in it
+	var/obj/item/grenade/grenade //You can remove the stuffing from a plushie and add a grenade to it for *nefarious uses*
 
 /obj/item/toy/plush/Initialize()
 	. = ..()
 	AddComponent(/datum/component/squeak, squeak_override)
 
+/obj/item/toy/plush/Destroy()
+	QDEL_NULL(grenade)
+	. = ..()
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: return ..()

File: code/game/objects/items/plushes.dm [hunk 17]
```diff
@@ -7,14 +7,66 @@
...
+			return
+		if(!user.transferItemToLoc(I, src))
+			return
+		user.visible_message("<span class='warning'>[user] slides [grenade] into [src].</span>", \
+		"<span class='danger'>You slide [I] into [src].</span>")
+		grenade = I
+		var/turf/T = get_turf(user)
+		log_game("[key_name(user)] added a grenade ([I.name]) to [src] at [COORD(T)].")
+		return
+	. = ..()
```
Comment (Cyberboss) [inline code/game/objects/items/plushes.dm]: return ..()

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**